patroni/postgresql/validator.py

```diff
@@ -128,456 +135,370 @@ def transform(self, name: str, value: Optional[Any]) -> Optional[Any]:
-    'wal_level': (
-        Enum(90300, 90400, ('minimal', 'archive', 'hot_standby')),
-        Enum(90400, 90600, ('minimal', 'archive', 'hot_standby', 'logical')),
-        Enum(90600, None, ('minimal', 'replica', 'logical'))
-    ),
-    'wal_log_hints': Bool(90400, None),
-    'wal_receiver_create_temp_slot': Bool(130000, None),
-    'wal_receiver_status_interval': Integer(90300, None, 0, 2147483, 's'),
-    'wal_receiver_timeout': Integer(90300, None, 0, 2147483647, 'ms'),
-    'wal_recycle': Bool(120000, None),
-    'wal_retrieve_retry_interval': Integer(90500, None, 1, 2147483647, 'ms'),
-    'wal_sender_timeout': Integer(90300, None, 0, 2147483647, 'ms'),
-    'wal_skip_threshold': Integer(130000, None, 0, 2147483647, 'kB'),
-    'wal_sync_method': Enum(90300, None, ('fsync', 'fdatasync', 'open_sync', 'open_datasync')),
-    'wal_writer_delay': Integer(90300, None, 1, 10000, 'ms'),
-    'wal_writer_flush_after': Integer(90600, None, 0, 2147483647, '8kB'),
-    'work_mem': Integer(90300, None, 64, 2147483647, 'kB'),
-    'xmlbinary': Enum(90300, None, ('base64', 'hex')),
-    'xmloption': Enum(90300, None, ('content', 'document')),
-    'zero_damaged_pages': Bool(90300, None)
-})
-
-
-recovery_parameters = CaseInsensitiveDict({
-    'archive_cleanup_command': String(90300, None),
-    'pause_at_recovery_target': Bool(90300, 90500),
-    'primary_conninfo': String(90300, None),
-    'primary_slot_name': String(90400, None),
-    'promote_trigger_file': String(120000, None),
-    'recovery_end_command': String(90300, None),
-    'recovery_min_apply_delay': Integer(90400, None, 0, 2147483647, 'ms'),
-    'recovery_target': Enum(90400, None, ('immediate', '')),
-    'recovery_target_action': Enum(90500, None, ('pause', 'promote', 'shutdown')),
-    'recovery_target_inclusive': Bool(90300, None),
-    'recovery_target_lsn': String(100000, None),
-    'recovery_target_name': String(90400, None),
-    'recovery_target_time': String(90300, None),
-    'recovery_target_timeline': String(90300, None),
-    'recovery_target_xid': String(90300, None),
-    'restore_command': String(90300, None),
-    'standby_mode': Bool(90300, 120000),
-    'trigger_file': String(90300, 120000)
-})
+#  key  - parameter name
+#  value - tuple of tuples. Each sub-tuple represents a different validation of the GUC across postgres versions. If a
+#          GUC validation has never changed over time, then it will have a single sub-tuple. For example,
+#          `password_encryption` used to be a boolean GUC up to Postgres 10, at which point it started being an enum.
+#          In that case the value of `password_encryption` would be a tuple of 2 tuples, each one reprensenting a
+#          different validation rule.
+parameters = CaseInsensitiveDict()
+recovery_parameters = CaseInsensitiveDict()
+
+
+def _get_module() -> 'ModuleType':
+    """Get a reference to the current module.
+
+    :returns: the module object.
+    """
+    return sys.modules[__name__]
```

Comment:
It is used only to find references to:
1. child classes of `_Transformable`
2. `parameters` + `recovery_parameters`

Both could be just a static mapping.

Or, actually https://github.com/zalando/patroni/pull/2671/files#diff-c1f5f15afbae52f3272b8520d30aea67a3820c6855b92a90ed5494f27fdd1ec3R171 could be easily built  dynamically:
```python
class _Transformable(abc.ABC):
    @classmethod
    def get_subclasses(cls):
        for subclass in cls.__subclasses__():
            for subsubclass in subclass.get_subclasses():
                yield subsubclass
            yield subclass


class ValidatorFactory:
    TYPES = {cls.__name__: cls for cls in _Transformable.get_subclasses()}
```

Reply:
Good idea, thanks! I just submitted a commit covering that suggestion.